sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java

```diff
@@ -525,20 +561,31 @@ private void acquireChannel(final Promise<Channel> promise) {
                 for (Channel channel : this.availableChannels) {
 
                     final RntbdRequestManager manager = channel.pipeline().get(RntbdRequestManager.class);
-                    final long pendingRequestCount = manager.pendingRequestCount();
 
-                    if (pendingRequestCount < pendingRequestCountMin) {
-                        pendingRequestCountMin = pendingRequestCount;
-                        candidate = channel;
+                    if (manager == null) {
+                        logger.warn("Channel({}) closed", channel);
+                    } else {
+                        final long pendingRequestCount = manager.pendingRequestCount();
+                        if (isChannelServiceable(channel, true) && pendingRequestCount < pendingRequestCountMin) {
+                            pendingRequestCountMin = pendingRequestCount;
+                            candidate = channel;
+                        }
                     }
                 }
 
-                assert candidate != null;
-
-                this.availableChannels.remove(candidate);
-                doAcquireChannel(promise, candidate);
-
-                return;
+                if (candidate != null && this.availableChannels.remove(candidate)) {
+                    this.doAcquireChannel(promise, candidate);
+                    return;
+                }
+            } else {
+                for (Channel channel : this.availableChannels) {
```

Comment:
Concurrent select of a channel is fine right? 

My assumption is the channel usage is guarded by the EvetnLoop so even the same channel is given back multiple timeback then it should be okey right? 

With the same assumption and assuming that uneven load-balacing with-in limits is acceptable, then will a single queue and linear check of availability work?

Reply:
Added comment with explanation